plugins/mapper-hll/src/main/java/org/elasticsearch/index/mapper/hll/HLLFieldMapper.java

```diff
@@ -0,0 +1,249 @@
+            Builder builder = new Builder(name);
+
+            // tweaking these settings is no longer allowed, the entire purpose of murmur3 fields is to store a hash
+            if (node.get("doc_values") != null) {
+                throw new MapperParsingException("Setting [doc_values] cannot be modified for field [" + name + "]");
+            }
+            if (node.get("index") != null) {
+                throw new MapperParsingException("Setting [index] cannot be modified for field [" + name + "]");
+            }
+
+            if (parserContext.indexVersionCreated().before(Version.V_5_0_0_alpha2)) {
+                node.remove("precision_step");
+            }
+
+            TypeParsers.parseField(builder, name, node, parserContext);
+
+            return builder;
+        }
+    }
+
+    // this only exists so a check can be done to match the field type to using murmur3 hashing...
+    public static class Murmur3FieldType extends MappedFieldType {
+        public Murmur3FieldType() {
+        }
+
+        protected Murmur3FieldType(Murmur3FieldType ref) {
+            super(ref);
+        }
+
+        @Override
+        public String typeName() {
+            return CONTENT_TYPE;
+        }
+
+        @Override
+        public Murmur3FieldType clone() {
+            return new Murmur3FieldType(this);
+        }
+
+        @Override
+        public IndexFieldData.Builder fielddataBuilder(String fullyQualifiedIndexName) {
+            failIfNoDocValues();
+            return new BytesBinaryDVIndexFieldData.Builder();
+        }
+
+        @Override
+        public Query existsQuery(QueryShardContext context) {
+            return new DocValuesFieldExistsQuery(name());
+        }
+
+        @Override
+        public Query termQuery(Object value, QueryShardContext context) {
+            throw new QueryShardException(context, "HLL fields are not searchable: [" + name() + "]");
+        }
+    }
+
+    protected HLLFieldMapper(String simpleName, MappedFieldType fieldType, MappedFieldType defaultFieldType,
+            Settings indexSettings, MultiFields multiFields, CopyTo copyTo) {
+        super(simpleName, fieldType, defaultFieldType, indexSettings, multiFields, copyTo);
+    }
+
+    @Override
+    protected String contentType() {
+        return CONTENT_TYPE;
+    }
+
+    @Override
+    protected void parseCreateField(ParseContext context, List<IndexableField> fields)
+            throws IOException {
+        // TODO: how do we take multiple values and HLL them?
+        // Ideas:
+        // - do we need to use parse() rather than parseCreateField()?
+        // - should we copy GeoPointFieldMapper, since that impl uses parse() and START_ARRAY and similar?
+
+        final Object value;
+
+        // try to parse it as a map
+        XContentParser.Token token = context.parser().currentToken();
+        if (token == XContentParser.Token.START_OBJECT) {
+            Map<String, Object> map = context.parser().map();
```

Comment:
Should be aware that this probably allocates the whole `Map` into memory, and, in practice, we're expecting this to be a pretty large map. It seems like the `ParseContext` object is flexible enough to support streaming parsing of JSON, so, the right way to handle that would be to advance, token at a time, via calls to `.parser().currentToken()` and `parser().nextToken()`, and actually offer the items to the in-memory HLL one-string-at-a-time as you come across the `VALUE_STRING` objects inside the array.